// react-vis/dist/lib/plot/axis-ticks.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _axisUtils = require('../utils/axis-utils');

var _scalesUtils = require('../utils/scales-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } // Copyright (c) 2016 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var LEFT = _axisUtils.ORIENTATION.LEFT;
var RIGHT = _axisUtils.ORIENTATION.RIGHT;
var TOP = _axisUtils.ORIENTATION.TOP;
var BOTTOM = _axisUtils.ORIENTATION.BOTTOM;


var propTypes = {
  width: _react2.default.PropTypes.number.isRequired,
  height: _react2.default.PropTypes.number.isRequired,
  orientation: _react2.default.PropTypes.oneOf([LEFT, RIGHT, TOP, BOTTOM]).isRequired
};

function _getTickTextAttributes(orientation) {
  var textAnchor = orientation === LEFT ? 'end' : orientation === RIGHT ? 'start' : 'middle';
  var dy = orientation === TOP ? '0' : orientation === BOTTOM ? '0.72em' : '0.32em';
  return {
    textAnchor: textAnchor,
    dy: dy
  };
}

var AxisTicks = function (_React$Component) {
  _inherits(AxisTicks, _React$Component);

  function AxisTicks(props) {
    _classCallCheck(this, AxisTicks);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(AxisTicks).call(this, props));
  }

  _createClass(AxisTicks, [{
    key: '_getTickFormatFn',
    value: function _getTickFormatFn(scale) {
      var tickFormat = this.props.tickFormat;

      return !tickFormat ? scale.tickFormat ? scale.tickFormat() : function (v) {
        return v;
      } : tickFormat;
    }
  }, {
    key: '_getTickValues',
    value: function _getTickValues(scale) {
      var _props = this.props;
      var tickTotal = _props.tickTotal;
      var tickValues = _props.tickValues;

      return !tickValues ? scale.ticks ? scale.ticks(tickTotal) : scale.domain() : tickValues;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props;
      var orientation = _props2.orientation;
      var width = _props2.width;
      var height = _props2.height;

      var x = orientation === LEFT ? width : 0;
      var y = orientation === TOP ? height : 0;

      var _props3 = this.props;
      var attr = _props3.attr;
      var tickSize = _props3.tickSize;
      var tickPadding = _props3.tickPadding;

      var isVertical = orientation === LEFT || orientation === RIGHT;
      var _props4 = this.props;
      var _props4$tickSizeInner = _props4.tickSizeInner;
      var tickSizeInner = _props4$tickSizeInner === undefined ? tickSize : _props4$tickSizeInner;
      var _props4$tickSizeOuter = _props4.tickSizeOuter;
      var tickSizeOuter = _props4$tickSizeOuter === undefined ? tickSize : _props4$tickSizeOuter;

      var scale = (0, _scalesUtils.getAttributeScale)(this.props, attr);

      var wrap = orientation === LEFT || orientation === TOP ? -1 : 1;

      var values = this._getTickValues(scale);
      var tickFormatFn = this._getTickFormatFn(scale);

      var tickXAttr = isVertical ? 'y' : 'x';
      var tickYAttr = isVertical ? 'x' : 'y';

      var ticks = values.map(function (v, i) {
        var _pathProps, _extends2;

        var pos = scale(v);
        var text = tickFormatFn(v);
        var pathProps = (_pathProps = {}, _defineProperty(_pathProps, tickXAttr + '1', pos), _defineProperty(_pathProps, tickXAttr + '2', pos), _defineProperty(_pathProps, tickYAttr + '1', -wrap * tickSizeOuter), _defineProperty(_pathProps, tickYAttr + '2', wrap * tickSizeInner), _pathProps);
        var textProps = _extends((_extends2 = {}, _defineProperty(_extends2, tickXAttr, pos), _defineProperty(_extends2, tickYAttr, wrap * (tickSizeOuter + tickPadding)), _extends2), _getTickTextAttributes(orientation));
        return _react2.default.createElement(
          'g',
          { key: i, className: 'rv-xy-plot__axis__tick' },
          _react2.default.createElement('line', _extends({}, pathProps, { className: 'rv-xy-plot__axis__tick__line' })),
          _react2.default.createElement(
            'text',
            _extends({}, textProps, { className: 'rv-xy-plot__axis__tick__text' }),
            text
          )
        );
      });

      return _react2.default.createElement(
        'g',
        {
          transform: 'translate(' + x + ', ' + y + ')',
          className: 'rv-xy-plot__axis__ticks' },
        ticks
      );
    }
  }]);

  return AxisTicks;
}(_react2.default.Component);

AxisTicks.displayName = 'AxisTicks';
AxisTicks.propTypes = propTypes;

exports.default = AxisTicks;
