// react-vis/dist/lib/legends/continous-color-legend.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _discreteColorLegendItem = require('./discrete-color-legend-item');

var _discreteColorLegendItem2 = _interopRequireDefault(_discreteColorLegendItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2016 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var RequiredValueType = _react2.default.PropTypes.oneOf([_react2.default.PropTypes.number, _react2.default.PropTypes.string]).isRequired;

var propTypes = {
  startColor: RequiredValueType,
  endColor: RequiredValueType,
  startTitle: RequiredValueType,
  endTitle: RequiredValueType,
  height: _react2.default.PropTypes.number,
  width: _react2.default.PropTypes.number
};

function ContinousColorLegend(_ref) {
  var startColor = _ref.startColor;
  var endColor = _ref.endColor;
  var startTitle = _ref.startTitle;
  var endTitle = _ref.endTitle;
  var height = _ref.height;
  var width = _ref.width;

  return _react2.default.createElement('div', { className: 'rv-continuous-color-legend', style: { width: width, height: height } });
}

ContinousColorLegend.displayName = 'ContinousColorLegend';
ContinousColorLegend.propTypes = propTypes;

exports.default = ContinousColorLegend;
