// react-vis/dist/lib/legends/search-wrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  searchText: _react2.default.PropTypes.string,
  onSearchChange: _react2.default.PropTypes.func,
  searchPlaceholder: _react2.default.PropTypes.string,
  searchFn: _react2.default.PropTypes.function.isRequired,
  width: _react2.default.PropTypes.number,
  height: _react2.default.PropTypes.number
}; // Copyright (c) 2016 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var defaultProps = {
  searchText: '',
  searchFn: function searchFn(items, s) {
    return items.filter(function (_ref) {
      var title = _ref.title;
      return title.toLowerCase().indexOf(s) !== -1;
    });
  }
};

function SearchWrapper(_ref2) {
  var children = _ref2.children;
  var onSearchChange = _ref2.onSearchChange;
  var searchText = _ref2.searchText;
  var searchPlaceholder = _ref2.searchPlaceholder;
  var width = _ref2.width;
  var height = _ref2.height;

  var onChange = onSearchChange ? function (_ref3) {
    var value = _ref3.target.value;
    return onSearchChange(value);
  } : null;
  return _react2.default.createElement(
    'div',
    { className: 'rv-search-wrapper', style: { width: width, height: height } },
    _react2.default.createElement(
      'form',
      { className: 'rv-search-wrapper__form' },
      _react2.default.createElement('input', {
        type: 'search',
        placeholder: searchPlaceholder,
        className: 'rv-search-wrapper__form__input',
        value: searchText,
        onChange: onChange })
    ),
    _react2.default.createElement(
      'div',
      { className: 'rv-search-wrapper__contents' },
      children
    )
  );
}

SearchWrapper.propTypes = propTypes;
SearchWrapper.defaultProps = defaultProps;
SearchWrapper.displayName = 'SearchWrapper';

exports.default = SearchWrapper;
